### ISSUE TITLE
Need some help to integrate a gateway ...

### ISSUE THREAD

**Issue by scramatte** [gr8shivam/laravel-sms-api#8]
Hi,

I need some help to integrate another gateway that requires following JSON payload.

```
{
    "api_key":"399d2b438a53ebed3db8a7d52107f846",
    "report_url":"http://yourserver.com/callback/script",
    "concat":1,
    "messages":[
        {
            "from":"GOOD PIZZA",
            "to":"34666666111",
            "text":"Hi John, today 2x1 in pizzas, watch the game like a boss with our new pepperoni pizza!",
            "send_at":"2018-02-18 17:30:00"
        }
    ]
}
```
I've create following code but something goes wrong. Can you give me an hand?

```
 'gateway360' => [
        'method' => 'POST', //Choose Request Method (GET/POST)
        'url' => 'https://api.gateway360.com/api/3.0/sms/send', //Base URL
        'params' => [
            'send_to_param_name' => '', //Send to Parameter Name
            'msg_param_name' => '', //Message Parameter Name
            'others' => [
                'api_key' => env('SMS_API_KEY'),
                'concat' => '1',
                'messages' => [
                        'from' => env('SMS_API_SENDER'),
                        'to' => 'to',
                        'text' => 'message'
                ]
            ],
        ],
        'json' => true, // Use if you want the params to be sent in JSON format instead of query params
        'add_code' => false, //Include Country Code (true/false)
    ]

```

**Comment by scramatte** [comment]
I've been able to send a SMS but  I'm unable to pickup to and message variables.

**Comment by scramatte** [comment]
Ok, the problem is that the class doesn't let to use any  structure for JSON payload.

**Comment by gr8shivam** [comment]
Hi @scramatte, we've currently 2 types of JSON payload structures which are most popular.

Structure 1: Sends everything in a single object

```
{
    "param1":"val1",
    "param2":"val2",
    "to":"Receipient1", //Also accepts array of Recipients
    "message":"Message",
}
```

Structure 2: Adds a wrapper for message & to in the format:

```
{
    "param1":"val1",
    "param2":"val2",
    "wrapper_name": [
    {
      "message": "Message",
      "to": [
        "Receipient1",
        "Receipient2"
      ]
    }
  ]
}
```

**Comment by gr8shivam** [comment]
Yes you got that right, the plugin picks `send_to_param_name` for `to` and `msg_param_name` for `message`. So, if their position is changed, it doesn't work.

-----

> I need to have "to" and "message" variables inside "messages" block. Some mechanism to search and replace pseudo variables is needed such {{to}} / {{message}}.

I understand this issue of custom JSON payload structure, but couldn't find a way to resolve this. I will relook into this. Would really appreciate if you can help on this.

**Comment by gr8shivam** [comment]
I looked upon your gateway documentation and found that the plugin wrapper option solves the payload structure except a few variables. I will work upon this and push an update to the plugin.

Please let me know your email to contact you further in this regard.

**Comment by gr8shivam** [comment]
Hi @scramatte , I'm looking forward to hear from you regarding this issue. I've made the required updates to the plugin and require you to test it. Please let me know your email to share the same.

**Comment by gr8shivam** [comment]
Fixed using wrapper params introduced in v3.0.2.

Here's a sample config for your gateway:-
```
    'gateway360' => [
        'method' => 'POST',
        'url' => 'https://api.gateway360.com/api/3.0/sms/send',
        'params' => [
            'send_to_param_name' => 'to',
            'msg_param_name' => 'text',
            'others' => [
                'api_key' => 'YOUR_API_KEY', //Your auth key
                'report_url' => 'AS_PER_GATEWAY_REQUIREMENT', //As per requirement
                'concat' => '1',  //As per requirement
            ],
        ],
        'json' => true,
        'jsonToArray' => false,
        'wrapper' => 'messages',
        'wrapperParams' => [
            'from' => 'YOUR_SENDER_NAME', //Your Sender ID
        ],
        'add_code' => false, //Include Country Code (true/false)
    ],
```

NOTE: If the mobile numbers don't have country code added to them, change `add_code` to `true` and set country code in function calling as `smsapi()->countryCode('COUNTRY_CODE')->sendMessage("TO", "Message");` .

To add the `send_at ` wrapper parameter, you can use `smsapi()->addWrapperParams(['send_at'=>'DATE_TIME_VAL'])->sendMessage("TO", "Message")`
